_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,257 @@
+
+With time, this implementation showed a number of problems and limitations:
+
+1. It requires ZooKeeper for cluster management, so it is not compatible with using Strimzi in KRaft mode.
+2. It doesn't scale well as it operates on one topic at a time, and it has to update a persistent store containing topic metadata.
+3. Due to its internal complexity, there are corner cases which are still not well understood (invalid state store, topic recreation after delete)
+4. There is a race condition between the operator and concurrent external admin operations that is not properly handled.
+
+For these reasons, we decided to create a new implementation called **Unidirectional Topic Operator (UTO)**.
+The new operator is fully backwards compatible, but it will only reconcile topic changes in one direction, from Kubernetes to Kafka.
+Your custom resources will be the source-of-truth, like with the other operators.
+
+<!--more-->
+
+### Introduction
+
+The Unidirectional Topic Operator enable creation, update, and deletion of Kafka topics by interacting with `KafkaTopic` custom resources in Kubernetes.
+It will only manage Kafka topics associated with `KafkaTopic` resources, and won't interfere with topics managed independently within the Kafka cluster.
+It can be deployed standalone, or along with your Kafka cluster using the `Kafka` custom resource.
+
+<figure>
+    <img src="/assets/images/posts/2023-10-23-uto-interactions.png" height=440>
+    <figcaption><small>Fig 1. Unidirectional Topic Operator interactions.</small></figcaption>
+</figure>
+
+Users can create and dynamically change Kafka topic configurations by interacting with the `KafkaTopic` custom resources in real-time.
+A reconciliation loop compares the desired state in `KafkaTopic` resources with the current state of Kafka topics in Kafka.
+Actions are taken to align the actual topic configurations with the custom resource specifications.
+If the Kafka-side configuration is changed out-of-band (e.g. by using the Admin client, or Kafka shell scripts), those changes will eventually be reverted.
+
+The operator requires that a single `KafkaTopic` resource is used to manage a single topic in a single Kafka cluster.
+It detects cases where multiple resources are attempting to manage a Kafka topic using the same `spec.topicName`.
+Only the oldest `KafkaTopic` resource will be reconciled, while the other will fail with a `ResourceConflict` error.
+
+In addition to be able to pause reconciliations, the operator also provides a way to unmanage a Kafka topic using annotations.
+If the `KafkaTopic` is paused and we delete it, then it will be also deleted in Kafka.
+If the `KafkaTopic` is unmanaged and we delete it, then only the resource will be garbage collected by Kubernetes, but the topic will still exists in Kafka.
+
+As with previous implementation, we still do not support decreasing `spec.partitions` and changing `spec.replicas`.
```

Comment:
```suggestion
As with previous implementation, we still do not support decreasing `.spec.partitions` and changing `.spec.replicas`.
```

Also, Kafka does not support a decreasing number of partitions. I think we should make that more clear by rephrasing this.